---
title: "(Day3) 細胞間相互作用解析ワークショップ: Cell-cell Interaction Analysis Workshop"
author:
- name: 露崎 弘毅（Koki Tsuyuzaki）
  affiliation:
  - Laboratory for Bioinformatics Research, RIKEN Center for Biosystems Dynamics Research, Japan
  - PRESTO, Japan Science and Technology Agency, Japan
  email: k.t.the-answer@hotmail.co.jp
output:
  bookdown::html_document2:
    toc: true
    toc_float: true
    theme: cosmo
    code_folding: show
    code_download: true
link-citations: true
vignette: >
  %\VignetteIndexEntry{Cell-cell Interaction Analysis Workshop}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  warning = FALSE,
  message = FALSE,
  fig.height = 7,
  fig.width = 10
)
```
<img src="../inst/images/sctensor.png" width="105" height="120">
<img src="../inst/images/lrbasedbi.png" width="105" height="120">
<img src="../inst/images/rtensor.png" width="105" height="120">
<img src="../inst/images/nntensor.png" width="105" height="120">
<img src="../inst/images/delayedtensor.png" width="105" height="120">
<img src="../inst/images/mwtensor.png" width="105" height="120">

## Description

In this workshop (presented in Japanese), you will learn how to detect cell-cell interaction from single-cell RNA-sequencing (scRNA-Seq) data using scTensor package. Moreover, you will learn some tensor frameworks of R language and will be able to freely use and decompose user's original tensor data.

Expectation: In the first half, you will learn how to use SingleCellExperiment object, transform the scRNA-Seq data to higher-order tensor, and apply scTensor against it. In the latter half, you will learn some tensor-related R packages.

Pre-requisites: The course is aimed at PhD students, Master’s students, and third & fourth year undergraduate students. Some basic R knowledge is assumed - this is not an introduction to R course. If you are not familiar with the R statistical programming language it is compulsory that you work through an introductory R course before you attend this workshop.

### Participation

After the lecture, participants are expected to follow along the hands-on session. we highly recommend participants bringing your own laptop.

### _R_ / _Bioconductor_ / _CRAN_ packages used

The following R/Bioconductor packages will be explicitly used:

List any _R_ / _Bioconductor_ / _CRAN_ packages that will be explicitly covered.

* SingleCellExperiment (Bioconductor)
* AnnotationHub (Bioconductor)
* LRBaseDbi (Bioconductor)
* scTensor (Bioconductor)
* einsum (CRAN)
* rTensor (CRAN)
* nnTensor (CRAN)
* HDF5Array (Bioconductor)
* DelayedArray (Bioconductor)
* DelayedTensor (Bioconductor)
* mwTensor (CRAN)

### Time outline

| Activity                                                  | Time|
|-----------------------------------------------------------|-----|
| Cell-cell Interaction Analysis by scTensor                | 20m |
| Advanced Topics about Tensor Arithmetic and Decomposition | 30m |
| Q & A                                                     | 10m |

## SingleCellExperiment

```{r}
library("SingleCellExperiment")
library("scTensor")

# テストデータ読み込み
data(GermMale) # 発現量行列
data(labelGermMale) # 細胞型ラベル
data(tsneGermMale) # t-SNEの二次元座標

# SingleCellExperimentオブジェクト生成
sce <- SingleCellExperiment(assays=list(counts = GermMale))
reducedDims(sce) <- SimpleList(TSNE=tsneGermMale$Y)

# ユーザー独自の正規化関数（オプション）
CPMED <- function(input){
    libsize <- colSums(input)
    median(libsize) * t(t(input) / libsize)
}

# 正規化（オプション）
normcounts(sce) <- log10(CPMED(counts(sce)) + 1)
```

## AnnotationHub

```{r}
library("AnnotationHub")
library("LRBaseDbi")

ah <- AnnotationHub()
dbfile <- query(ah, c("LRBaseDb", c("Homo sapiens", "v002")))[[1]]
LRBase.Hsa.eg.db <- LRBaseDbi::LRBaseDb(dbfile)
```

## scTensor
### 実データ

```{r}
# 実データ
# 各種設定（関係する情報がmetadata(sce)に登録される
cellCellSetting(sce, LRBase.Hsa.eg.db, names(labelGermMale))

# ランク推定
rks <- cellCellRanks(sce, assayNames="normcounts")

# CCIテンソル分解
set.seed(1234)
cellCellDecomp(sce, ranks=rks$selected, assayNames="normcounts")

# HTMLレポート出力
# cellCellReport(sce, reducedDimNames="TSNE", assayNames="normcounts",
#     title="Cell-cell interaction within Germline_Male, GSE86146",
#     author="Koki Tsuyuzaki", html.open=TRUE, upper=2,
#     goenrich=TRUE, meshenrich=FALSE, reactomeenrich=FALSE,
#     doenrich=FALSE, ncgenrich=FALSE, dgnenrich=FALSE)
```

### 人工データ

```{r}
params <- newCCSParams()
getParam(params, "nGene")
getParam(params, "nCell")
getParam(params, "cciInfo")

out_sim <- cellCellSimulate(params)
str(out_sim$input)
str(out_sim$LR)
L <- out_sim$input[out_sim$LR[,"GENEID_L"], ] # 500genes × 150cells
R <- out_sim$input[out_sim$LR[,"GENEID_R"], ] # 500genes × 150cells
```

## Advanced Topics

### einsum

```{r}
library("einsum")
# einsumのデモ（for文のSugar Syntax）
CCItensor <- einsum::einsum('ij,kj->ikj', L, R)

# for文だと、これくらい複雑
CCItensor2 <- array(0, dim=c(nrow(L), nrow(R), ncol(L)))
for(i in 1:nrow(L)){
    for(j in 1:ncol(L)){
        for(k in 1:nrow(R)){
            CCItensor2[i,k,j] <- L[i,j] * R[k,j]
        }
    }
}
identical(CCItensor, CCItensor2)
```

### rTensor

```{r}
library("rTensor")
# rTensorオブジェクト
CCItensor <- as.tensor(CCItensor)
str(CCItensor@data) # 元の配列データ
# 各種rTensor関数が使えるようになる
rTensor::modeSum(CCItensor, m=1, drop=TRUE) # 特定のモードでの和
rTensor::cs_unfold(CCItensor, m=1) # テンソルを引き延ばす（行列化）
rTensor::fnorm(CCItensor) # フロベニウスノルム
```

### nnTensor

```{r}
library("nnTensor")
# nnTensor（scTensorのエンジン部分）
out_nn <- nnTensor::NTD(CCItensor, algorithm="KL", rank=c(3,4,5))
```

### HDF5Array, DelayedArray, and DelayedTensor

```{r}
library("HDF5Array")
library("DelayedArray")
library("DelayedTensor")
# DelayedTensor（大規模テンソルデータ用）
options(delayedtensor.sparse = FALSE) # Sparse mode off
options(delayedtensor.verbose = TRUE) # Verbose message
setHDF5DumpCompressionLevel(level=9L) # No compress in HDF5 file
setAutoBlockSize(size=1E+7) # このサイズまではメモリに載せる
getAutoBlockSize()

# DelayedTensor版einsum
darr_L <- DelayedArray(L)
darr_R <- DelayedArray(R)
darr_CCItensor <- DelayedTensor::einsum('ij,kj->ikj', darr_L, darr_R)

# DelayedTensor版modeSum
DelayedTensor::modeSum(darr_CCItensor, m=1, drop=TRUE) # 特定のモードでの和
DelayedTensor::cs_unfold(darr_CCItensor, m=1) # テンソルを引き延ばす（行列化）
DelayedTensor::fnorm(darr_CCItensor) # フロベニウスノルム
```

### mwTensor

```{r}
library("mwTensor")
# mwTensor
# バラバラのままテンソル分解するテンソルフレームワーク
Xs <- mwTensor::toyModel("coupled_CP_Easy")
params <- new("CoupledMWCAParams", Xs=Xs, viz=TRUE)
out_mw <- CoupledMWCA(params)
str(out_mw@common_factors)
str(out_mw@common_cores)
```

## Session Information

```{r}
sessionInfo()
```
